### PR TITLE
feat: wallet picker

### PR DESCRIPTION
**Pull request by shamilovtim**
### Context

- We want to prototype workflows and support users being able to store credentials on wallets outside of PFI client apps.
- Native wallet APIs don't exist **yet** on every platform

There were two options for supporting wallets.

### A: HTML DOM wallet picker
- Use a remotely hosted website to have a wallet picker in HTML. Visit that website and forward the SIOP params to it, and then from it, to the end wallet.

### B: Native wallet picker
- **B.1**: Host a json file of wallets and links to those walles. Ask implementers to GET the JSON file, loop through it and forward the SIOP params to that wallet.
- **B.2**: Provide native SDKs that will do all of the B.1. logic for the user. Render unstyled platform native UI for the wallet picker.

This PR implements B1 first and only. In summary, this PR allows a wallet to pull down a remotely hosted JSON file, loop through the wallets present in it, and forward the SIOP params to that wallet.

TODO: That wallet will then handle issuance, and return control back to DidPay.

### Implementation

- Created a "service" for Wallet HTTP and Linking API
- Add a wallet picker screen
- GET JSON file and loop through it
- Universal link to TBD Identity Wallet with SIOP forwarded params
- Add improvements to support this functionality in the IDVService
- Add a feature flag for the wallet picker to allow disabling the workflow 

### Video
https://github.com/TBD54566975/didpay/assets/3495974/fd613f7c-fe6a-4e2c-b195-6e36e45ec55b

### CODE REVIEW

File: pubspec.yaml [hunk 1]
```diff
@@ -41,12 +42,12 @@ dev_dependencies:
...
-            weight: 600
+    - family: IBMPlexMono
+      fonts:
+        - asset: assets/fonts/IBMPlexMono-Regular.ttf
+        - asset: assets/fonts/IBMPlexMono-Medium.ttf
+          weight: 500
+        - asset: assets/fonts/IBMPlexMono-Bold.ttf
+          weight: 700
+        - asset: assets/fonts/IBMPlexMono-SemiBold.ttf
+          weight: 600
```
Comment (wesbillman) [inline pubspec.yaml:53]: That sweet sweet yaml formatter ❤️

**Comment by ethan-tbd** [comment]
this is really cool!! by "wallet", we're talking about applications meant to store identity credentials?

File: lib/features/wallets/wallet_selection_page.dart [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+              leading: Container(
+                width: Grid.md,
+                height: Grid.md,
+                decoration: BoxDecoration(
+                    color: Theme.of(context).colorScheme.surface,
+                    shape: BoxShape.rectangle,
+                    borderRadius: BorderRadius.circular(Grid.xxs)),
+                child: Center(
+                  child: Text(
+                    '💳',
```
Comment (ethan-tbd) [inline lib/features/wallets/wallet_selection_page.dart:101]: nit: could we use a Material Icon here? maybe `Icon(Icons.wallet)`

File: lib/features/wallets/wallet_selection_page.dart [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+                        if (selectedWallet.value?.name == 'DidPay') {
+                          Navigator.of(context).push(
+                            MaterialPageRoute(builder: (context) {
+                              return PfiVerificationPage(pfi: pfi);
+                            }),
+                          );
+                        } else {
+                          final result = await DidResolver.resolve(pfi.didUri);
+                          final widgetService = result.didDocument?.service
+                              ?.firstWhere((e) => e.type == 'IDV');
```
[the remark this answers is not in the document]
Comment (shamilovtim) [reply lib/features/wallets/wallet_selection_page.dart:58]: linter says that extension is unavailable

[the remark this answers is not in the document]
Comment (wesbillman) [reply lib/features/wallets/wallet_selection_page.dart:58]: You'll need `collection.dart`:
https://github.com/TBD54566975/didpay/blob/4eba15a20cc4ef61248d75fca214d0db4a9cb877/lib/features/pfis/pfi_confirmation_page.dart#L3

File: lib/features/onboarding/agreement_page.dart [hunk 4]
```diff
@@ -36,11 +38,17 @@ class AgreementPage extends HookWidget {
...
+                            if (FeatureFlags.hasWalletPicker) {
+                              return WalletSelectionPage(
+                                pfi: pfi,
+                              );
+                            } else {
+                              return PfiVerificationPage(
+                                pfi: pfi,
+                              );
+                            }
+                          }),
```
[the remark this answers is not in the document]
Comment (shamilovtim) [reply lib/features/onboarding/agreement_page.dart]: fixed 🫡

File: lib/services/wallet_service.dart [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import 'dart:convert';
+import 'package:didpay/features/wallets/wallet.dart';
+import 'package:http/http.dart' as http;
+
+class WalletService {
+  Future<List<Wallet>> getWallets() async {
+    final response =
+        await http.get(Uri.parse('https://tbd54566975.github.io/wallets.json'));
```
[the remark this answers is not in the document]
Comment (shamilovtim) [reply lib/services/wallet_service.dart]: fixed 🫡

File: lib/features/wallets/wallet_selection_page.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (shamilovtim) [reply lib/features/wallets/wallet_selection_page.dart]: fixed 🫡

File: lib/features/wallets/wallet_selection_page.dart [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+              ),
+              trailing: (selectedWalletState.value?.name == wallet.name)
+                  ? Icon(
+                      Icons.check,
+                      color: Theme.of(context).colorScheme.primary,
+                    )
+                  : null,
+              onTap: () {
+                selectedWalletState.value = wallet;
+                if (selectedWalletState.value?.name != 'DidPay') {
```
[the remark this answers is not in the document]
Comment (shamilovtim) [reply lib/features/wallets/wallet_selection_page.dart]: fixed 🫡

File: lib/features/wallets/wallet_selection_page.dart (hunk 2 above)
Comment (shamilovtim) [reply lib/features/wallets/wallet_selection_page.dart]: fixed 🫡

File: lib/config/feature_flags.dart [hunk 7]
```diff
@@ -0,0 +1,3 @@
+class FeatureFlags {
+  static bool get hasWalletPicker => true;
+}
```
[the remark this answers is not in the document]
Comment (shamilovtim) [reply lib/config/feature_flags.dart]: fixed 🫡

**Comment by shamilovtim** [comment]
> this is really cool!! by "wallet", we're talking about applications meant to store identity credentials? also, are we assuming that the Web5 Identity Wallet will store both IDV credentials (which direct users to the [pre-auth flow](https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#name-pre-authorized-code-flow)) and KCC credentials (which direct users to the [auth flow](https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#name-authorization-code-flow))?

It's a good question. I think so? I'm unsure and prototyping this right now. Has anyone got the sumsub credential yet and stuck it in long term storage yet or is that the next step?

File: lib/features/wallets/wallet_selection_page.dart [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+
+  @override
+  Widget build(BuildContext context, WidgetRef ref) {
+    final wallets = ref.watch(walletsProvider);
+    final selectedWallet =
+        useState<Wallet?>(Wallet(name: 'DidPay', url: 'didpay://'));
+
+    return Scaffold(
+        appBar: AppBar(),
+        body: switch (wallets) {
```
[the remark this answers is not in the document]
Comment (shamilovtim) [reply lib/features/wallets/wallet_selection_page.dart:30]: I think the current syntax is pretty clean what's the benefit of the alternative?

File: lib/features/wallets/wallet_selection_page.dart [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+          AsyncError(:final error) => Center(child: Text('Error: $error')),
```
[the remark this answers is not in the document]
Comment (shamilovtim) [reply lib/features/wallets/wallet_selection_page.dart]: fixed 🫡

File: lib/features/wallets/wallet_selection_page.dart (hunk 8 above)
[the remark this answers is not in the document]
Comment (wesbillman) [reply lib/features/wallets/wallet_selection_page.dart:30]: Cool works for me too! Just wanted to point out other options there, but I'm good with either approach.

**Comment by ethan-tbd** [comment]
> > this is really cool!! by "wallet", we're talking about applications meant to store identity credentials? also, are we assuming that the Web5 Identity Wallet will store both IDV credentials (which direct users to the [pre-auth flow](https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#name-pre-authorized-code-flow)) and KCC credentials (which direct users to the [auth flow](https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#name-authorization-code-flow))?
> 
> It's a good question. I think so? I'm unsure and prototyping this right now. Has anyone got the sumsub credential yet and stuck it in long term storage yet or is that the next step?

not yet, that's the next step!

**Review by ethan-tbd — APPROVED** [review]

**Comment by shamilovtim** [comment]
Someone else will need to merge as I don't have rights to do it

**Pull request merged**